### PR TITLE
Update the AWS Lambda deployment guide

### PR DESCRIPTION
**Pull request by rkoron007**

### CODE REVIEW

File: docs/source/deployment/lambda.mdx [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+> ⚠️ AWS best practices warn against using your AWS account root user keys for any task where it's not required (e.g., don't use these keys to configure the AWS CLI). Instead, [create an IAM user account](https://docs.aws.amazon.com/IAM/latest/UserGuide/getting-started_create-admin-group.html#getting-started_create-admin-group-console) with the [least privilege](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#grant-least-privilege) required to deploy your application, and [configure the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-config) to use that account.
+
+## Setting up your project
+
+For this example, we'll start from scratch to show how all the pieces fit together.
+
+Begin by installing the necessary packages for using Apollo Server and its integration for AWS Lambda:
+
+```shell
+npm install @apollo/server graphql @as-integrations/aws-lambda
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:28]: Does this TypeScript example work without installing `typescript` and explicitly building it somehow (eg does `serverless` magically bundle TypeScript for you or something?)

File: docs/source/deployment/lambda.mdx [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+      - http:
+          path: /
+          method: get
+          cors: true
+# Omit the below lines if you aren't using TS!
+plugins:
+  - serverless-plugin-typescript
+```
+
+> If you are using webpack, you might want to install and configure Serverless using the [`serverless-webpack` plugin](https://github.com/serverless-heaven/serverless-webpack) to ensure the CLI can find your file.
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:196]: I don't really know what this means. If this is really designed as a tutorial for people new to all of this, would you be using webpack if you haven't been told to?

File: docs/source/deployment/lambda.mdx [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+    events:
+      - http:
+          path: /
+          method: post
+          cors: true
+      - http:
+          path: /
+          method: get
+          cors: true
+# Omit the below lines if you aren't using TS!
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:191]:
```suggestion
# Omit the following lines if you aren't using TS!
```

File: docs/source/deployment/lambda.mdx [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+  name: aws
+  runtime: nodejs16.x
+functions:
+  graphql:
+    # Make sure your file path is correct!
+    # (e.g., if your file is in the root folder use server.graphqlHandler )
+    # The format is: <FILENAME>.<HANDLER>
+    handler: src/server.graphqlHandler # highlight-line
+    events:
+      - http:
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:183]: If I understand [the docs](https://www.serverless.com/framework/docs/providers/aws/guide/serverless.yml#api-gateway-v2-http-api-1) correctly, this is enabling API Gateway v1. Maybe makes sense to do v2 instead? Not an expert on the difference, just assume the newer one is better.

File: docs/source/deployment/lambda.mdx [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+If you ever want to remove the S3 bucket or Lambda functions that Serverless created, you can run the following command:
+
+```bash
+ serverless remove
+```
+
+## Customizing HTTP behavior
+
+The `@as-integrations/aws-lambda` package is compatible with Lambda's API Gateway V1 (REST) and V2 (HTTP). However, it does not let you customize HTTP behavior directly.
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:285]: It also does not support other AWS products that invoke Lambda functions, like Application Load Balancers.

File: docs/source/deployment/lambda.mdx [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+
+If you'd like to customize your HTTP behavior, you can couple Apollo Server's Express integration (i.e., [`expressMiddleware`](../api/express-middleware)) with the [`@vendia/serverless-express`](https://github.com/vendia/serverless-express) package. The `@vendia/serverless-express` library translates between Lambda events and Express requests.
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:287]: I think a parenthetical like "(Note that despite the name, this library is not related to the `serverless` CLI.)" or something would be helpful.

File: docs/source/deployment/lambda.mdx [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+```ts
+const server = new ApolloServer<MyContext>({
+  typeDefs,
+  resolvers,
+});
+
+// This final export is important!
+export const graphqlHandler = startServerAndCreateLambdaHandler(server, {
+  // highlight-start
+  context: async (event, context) => {
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:333]: I believe this takes `event` and `context` as options on a single argument.

File: docs/source/deployment/lambda.mdx [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+  },
+  // highlight-end
+});
+```
+
+</MultiCodeBlock>
+
+The `event` object contains the API Gateway event (HTTP headers, HTTP method, body, path, etc.). The `context` object (not to be confused with the `context` function) contains the current Lambda Context (function name, function version, `awsRequestId`, time remaining, etc.).
+
+If you've [changed your set up to use `@vendia/serverless-express`](#customizing-http-behavior) you'll also have access to an `express` argument, containing the Express request and response in your `context` function:
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:347]: I'm not sure what you mean by "an `express` argument" here.

If you're using this setup then the context function receives Express-style information as its argument, and the Lambda stuff is available via calling getCurrentInvoke. There's no argument named `express`.

File: docs/source/deployment/lambda.mdx [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+      };
+    },
+  }),
+);
+
+exports.handler = serverlessExpress({ app });
+```
+
+</MultiCodeBlock>
+````
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:387]:
```suggestion
```

File: docs/source/config.json [hunk 10]
```diff
@@ -45,7 +45,8 @@
       "Proxy configuration": "/security/proxy-configuration"
     },
     "Deployment": {
-      "Heroku": "/deployment/heroku"
+      "Heroku": "/deployment/heroku",
+      "Lambda": "/deployment/lambda"
```
Comment (glasser) [inline docs/source/config.json:49]: Perhaps put it first?

**Review by glasser — COMMENTED** [review]
Looks basically good. I did not actually try to walk through the tutorial steps myself though.

File: docs/source/deployment/lambda.mdx (hunk 2 above)
Comment (rkoron007) [reply docs/source/deployment/lambda.mdx:196]: I was thinking this might be a helpful callout for folks who are using webpack and might not have a clear file path. This was originally [mentioned in the PR adding info to the Lambda article](https://github.com/apollographql/apollo-server/pull/5854/files), and I figured I'd toss it in just in case someone using webpack was running into problems.

I can remove it if you don't feel like it's valuable though!

Comment (glasser) [reply docs/source/deployment/lambda.mdx:196]: I see. I know @michelfloyd's suggestion in #5854 is a year old (oops) but maybe they have more details here? I just think it might be a bit weird to, partway through a step-by-step tutorial, start talking about a different package without an explanation for why you would be adding that package to your build.

File: docs/source/deployment/lambda.mdx (hunk 9 above)
Comment (rkoron007) [reply docs/source/deployment/lambda.mdx:387]: Ooops, thank you!

File: docs/source/deployment/lambda.mdx (hunk 2 above)
Comment (rkoron007) [reply docs/source/deployment/lambda.mdx:196]: That's a really good point, I'll remove this callout it does feel kind of weird looking at it again.

File: docs/source/deployment/lambda.mdx (hunk 4 above)
Comment (rkoron007) [reply docs/source/deployment/lambda.mdx:183]: Great catch! Updated to using v2:

```yaml title="serverless.yml"
service: apollo-lambda
provider:
  name: aws
  runtime: nodejs16.x
  httpApi:
    cors: true
functions:
  graphql:
    # Make sure your file path is correct!
    # (e.g., if your file is in the root folder use server.graphqlHandler )
    # The format is: <FILENAME>.<HANDLER>
    handler: src/server.graphqlHandler # highlight-line
    events:
      - httpApi:
          path: /
          method: POST
      - httpApi:
          path: /
          method: GET
# Omit the following lines if you aren't using TS!
plugins:
  - serverless-plugin-typescript
```

File: docs/source/deployment/lambda.mdx (hunk 5 above)
Comment (rkoron007) [reply docs/source/deployment/lambda.mdx:285]: Nice call out, tweaked this to: 

> The `@as-integrations/aws-lambda` package is compatible with Lambda's API Gateway V1 (REST) and V2 (HTTP). However, it does not let you customize HTTP behavior directly or support other AWS products that invoke Lambda functions (e.g., such as Application Load Balancers).

File: docs/source/deployment/lambda.mdx (hunk 6 above)
Comment (rkoron007) [reply docs/source/deployment/lambda.mdx:287]: Added an additional sentence:

> Despite their similar names, the Serverless CLI and the `@vendia/serverless-express` package are unrelated.

File: docs/source/deployment/lambda.mdx (hunk 7 above)
Comment (rkoron007) [reply docs/source/deployment/lambda.mdx:333]: Oh, thank you! Great catch 🔥

File: docs/source/deployment/lambda.mdx (hunk 8 above)
Comment (rkoron007) [reply docs/source/deployment/lambda.mdx:347]: I forgot to update that sentence, thank you! I changed it to:

> If you've [changed your setup to use `@vendia/serverless-express`](#customizing-http-behavior) your `context` function receives `req` and `res` options which are [`express.Request`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2c4c1d0b106a1305c384cde84bb14bbf1a938d2c/types/express-serve-static-core/index.d.ts#L367) and [`express.Response`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2c4c1d0b106a1305c384cde84bb14bbf1a938d2c/types/express-serve-static-core/index.d.ts#L659) objects:

As well as updated the code snippet with a callout for where folks have access to the API Gateway and Lambda context:
```ts
app.use(
  cors(),
  json(),
  expressMiddleware(server, {
    // The Express request and response objects are passed into
    // your context initialization function
    context: async ({ req, res }) => {
      // Here is where you'll have access to the
      // API Gateway event and Lambda Context
      const { event, context } = serverlessExpress.getCurrentInvoke();
      return {
        expressRequest: req,
        expressResponse: res,
        lambdaEvent: event,
        lambdaContext: context,
      };
    },
  }),
);
```

File: docs/source/deployment/lambda.mdx [hunk 11]
```diff
@@ -282,9 +280,9 @@ If you ever want to remove the S3 bucket or Lambda functions that Serverless cre
 
 ## Customizing HTTP behavior
 
-The `@as-integrations/aws-lambda` package is compatible with Lambda's API Gateway V1 (REST) and V2 (HTTP). However, it does not let you customize HTTP behavior directly.
+The `@as-integrations/aws-lambda` package is compatible with Lambda's API Gateway V1 (REST) and V2 (HTTP). However, it does not let you customize HTTP behavior directly or support other AWS products that invoke Lambda functions (e.g., such as Application Load Balancers).
```
Comment (glasser) [inline docs/source/deployment/lambda.mdx:283]: "eg" and "such as" are redundant

**Review by glasser — APPROVED** [review]
caveat: have not actually tried working through the tutorial myself
